### PR TITLE
objstore : implement Baidu BOS

### PR DESCRIPTION
**Pull request by yahaa**
Try to reimplement Baidu BOS objstore.

Reference: https://github.com/thanos-io/thanos/pull/1329

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Verification

```bash
# prepare some envs
$ export BOS_BUCKET=your_bos_bucket BOS_ENDPOINT=endpoint BOS_ACCESS_KEY=your_baidu_account_access_key BOS_SECRET_KEY=your_baidu_account_secret_key THANOS_ALLOW_EXISTING_BUCKET_USE=true THANOS_TEST_OBJSTORE_SKIP=GCS,S3,AZURE,SWIFT,COS,ALIYUNOSS

# run e2etest
$  go test -v ./pkg/objstore/... 
```

### CODE REVIEW

**Comment by yahaa** [comment]
@bwplotka @jojohappy Please help review this PR, thanks!

File: pkg/objstore/bos/bos.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+}
+
+// Name returns the bucket name for the provider.
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(_ context.Context, name string) error {
+	if err := b.client.DeleteObject(b.name, name); err != nil {
```
Comment (GiedriusS) [inline pkg/objstore/bos/bos.go:104]: Just do: `return b.client.DeleteObject(b.name, name)`

File: pkg/objstore/bos/bos.go [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+		logger = log.NewNopLogger()
+	}
+
+	config, err := parseConfig(conf)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing bos configuration")
+	}
+
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "validate bos configuration")
```
Comment (GiedriusS) [inline pkg/objstore/bos/bos.go:79]: All error messages need to have the form of `base verb + present participle $something`. So, it should be: `validating BOS configuration`. This way it looks nice when a couple errors are stringed together:
`creating bucket: validating BOS configuration: insufficient bos configuration information`

File: pkg/objstore/bos/bos.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+		return err
+	}
+	return nil
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(_ context.Context, name string, r io.Reader) error {
+	size, err := objstore.TryToGetSize(r)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get size apriori to upload %s", name)
```
Comment (GiedriusS) [inline pkg/objstore/bos/bos.go:114]: Same here: `errors.Wrapf(err, "getting size of %s", name)`

File: pkg/objstore/bos/bos.go [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+	var marker string
+	for {
+		if err := ctx.Err(); err != nil {
+			return errors.Wrap(err, "context closed while iterating bucket")
+		}
+
+		objects, err := b.client.ListObjects(b.name, &api.ListObjectsArgs{
+			Delimiter: delimiter,
+			Marker:    marker,
+			MaxKeys:   1000,
```
Comment (GiedriusS) [inline pkg/objstore/bos/bos.go:198]: What if there are more than that? :thinking:

File: pkg/objstore/bos/bos.go [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+		marker = objects.NextMarker
+		for _, object := range objects.Contents {
+			if err := f(object.Key); err != nil {
+				return errors.Wrapf(err, "callback func invoke for object %s failed ", object.Key)
+			}
+		}
+
+		for _, object := range objects.CommonPrefixes {
+			if err := f(object.Prefix); err != nil {
+				return errors.Wrapf(err, "callback func invoke for object %s failed ", object.Prefix)
```
Comment (GiedriusS) [inline pkg/objstore/bos/bos.go:215]: Spurious blank line and the end here + please update the error messages according to the form in the previous comment.

File: pkg/objstore/bos/bos_test.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"bytes"
+	"context"
+	"os"
+	"testing"
+
+	"github.com/thanos-io/thanos/pkg/testutil"
+)
+
+func TestBOSUpload(t *testing.T) {
```
Comment (GiedriusS) [inline pkg/objstore/bos/bos_test.go:15]: I think it is already covered by e2e tests :) that's why you have the NewTestBucket() function. If you have updated `pkg/objstore/objtesting/foreach.go` then it should be possible to run the test suite via that. Does it not work for some reason?

**Review by GiedriusS — COMMENTED** [review]
There's no context.Context support? :(

File: pkg/objstore/bos/bos.go (hunk 4 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go:195]: > There's no context.Context support? :(

Yes, not supported in current SDK.

File: pkg/objstore/bos/bos.go (hunk 1 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go]: Done.

File: pkg/objstore/bos/bos.go (hunk 2 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go]: Thanks, I have been revised, please review again.

File: pkg/objstore/bos/bos.go (hunk 5 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go]: Thanks, I have been revised, please review again.

File: pkg/objstore/bos/bos.go (hunk 4 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go:195]: > What if there are more than that? 🤔

If there are more than 1000, the server will return objects with two fields `objects.NextMarker` and `objects.IsTruncated`, if  !objects.IsTruncated, we continue to ListObjects in the next loop.

File: pkg/objstore/bos/bos_test.go (hunk 6 above)
Comment (yahaa) [reply pkg/objstore/bos/bos_test.go:15]: I'm not sure whether e2e tests can cover that case where the obj are very large, so i do it.

File: pkg/objstore/bos/bos.go (hunk 4 above)
Comment (bwplotka) [reply pkg/objstore/bos/bos.go:195]: Can we add at least issue on SDK on ability to cancel request from context? 🤗

File: pkg/objstore/bos/bos_test.go (hunk 6 above)
Comment (bwplotka) [reply pkg/objstore/bos/bos_test.go:15]: Can we then add such test case to `AcceptanceTest` function in  `pkg/objstore/testing.go:82`? 🤗

File: pkg/objstore/objtesting/foreach.go [hunk 7]
```diff
@@ -148,4 +149,17 @@ func ForeachStore(t *testing.T, testFn func(t *testing.T, bkt objstore.Bucket))
 			testFn(t, bkt)
 		})
 	}
+
+	// Optional BOS.
+	if !IsObjStoreSkipped(t, client.BOS) {
+		t.Run("Baidu bos", func(t *testing.T) {
```
Comment (bwplotka) [inline pkg/objstore/objtesting/foreach.go:155]:
```suggestion
		t.Run("Baidu BOS", func(t *testing.T) {
```

File: pkg/testutil/testutil.go [hunk 8]
```diff
@@ -173,6 +173,8 @@ func TolerantVerifyLeakMain(m *testing.M) {
 		// https://github.com/kubernetes/klog/blob/c85d02d1c76a9ebafa81eb6d35c980734f2c4727/klog.go#L417
 		goleak.IgnoreTopFunction("k8s.io/klog/v2.(*loggingT).flushDaemon"),
 		goleak.IgnoreTopFunction("k8s.io/klog.(*loggingT).flushDaemon"),
+		// https://github.com/baidubce/bce-sdk-go/blob/9a8c1139e6a3ad23080b9b8c51dec88df8ce3cda/util/log/logger.go#L359
```
Comment (bwplotka) [inline pkg/testutil/testutil.go:176]: Can we add issue for this on github.com/baidubce/bce-sdk-go?

**Review by bwplotka — COMMENTED** [review]
Thanks for this! Some comments, but generally looks amazing!

File: pkg/objstore/objtesting/foreach.go (hunk 7 above)
Comment (yahaa) [reply pkg/objstore/objtesting/foreach.go]: Ok, Thanks!

File: pkg/objstore/bos/bos_test.go (hunk 6 above)
Comment (yahaa) [reply pkg/objstore/bos/bos_test.go]: Thanks for your advice, I have added this case to AcceptanceTest.

File: pkg/testutil/testutil.go (hunk 8 above)
Comment (yahaa) [reply pkg/testutil/testutil.go:176]: I have reported this issue with our internal colleagues, and I will continue to follow up the SDK update.

File: pkg/objstore/bos/bos.go (hunk 4 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go:195]: There is already an issue about this. 😂

File: pkg/testutil/testutil.go (hunk 8 above)
Comment (bwplotka) [reply pkg/testutil/testutil.go:176]: 💪🏽

File: docs/storage.md [hunk 9]
```diff
@@ -418,6 +418,23 @@ config:
 
 Use --objstore.config-file to reference to this configuration file.
 
+#### Baidu BOS
+
+In order to use Baidu BOS object storage, you should apply a Baidu Account to create an object storage bucket at first. Note that detailed from Baidu Cloud Documents:[https://cloud.baidu.com/doc/BOS/index.html](https://cloud.baidu.com/doc/BOS/index.html) for more detail.
```
Comment (bwplotka) [inline docs/storage.md:423]:
```suggestion
In order to use Baidu BOS object storage, you should apply for a Baidu Account and create an object storage bucket first. Refer to [Baidu Cloud Documents](https://cloud.baidu.com/doc/BOS/index.html) for more details.
```

File: docs/storage.md [hunk 10]
```diff
@@ -418,6 +418,23 @@ config:
...
+
+To use Baidu BOS object storage, please specify following yaml configuration file in `objstore.config*` flag.
```
Comment (bwplotka) [inline docs/storage.md:425]:
```suggestion
To use Baidu BOS object storage, please specify the following yaml configuration file in `--objstore.config*` flag.
```

File: docs/storage.md [hunk 11]
```diff
@@ -418,6 +418,23 @@ config:
...
+type: BOS
+config:
+  bucket: ""
+  endpoint: ""
+  access_key: ""
+  secret_key: ""
+```
+
+Use --objstore.config-file to reference to this configuration file.
+
```
Comment (bwplotka) [inline docs/storage.md:437]:
```suggestion
```

Not sure what this sentenced tried to say. I think it's fine whatever we have above

File: pkg/objstore/bos/bos.go [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+			return err
+		}
+
+		objects, err := b.client.ListObjects(b.name, &api.ListObjectsArgs{
+			Delimiter: delimiter,
+			Marker:    marker,
+			MaxKeys:   1000,
+			Prefix:    dir,
+		})
+
```
Comment (bwplotka) [inline pkg/objstore/bos/bos.go:198]:
```suggestion
```

nit: It's useful to bring err and err check together without a new line.

**Review by bwplotka — APPROVED** [review]
Amazing. LGTM, only suggestion towards documentation nits. 

Thanks!

File: pkg/objstore/bos/bos.go (hunk 12 above)
Comment (yahaa) [reply pkg/objstore/bos/bos.go]: > Amazing. LGTM, only suggestion towards documentation nits.
> 
> Thanks!

Ok thanks, I have updated the documentation.

File: docs/storage.md (hunk 10 above)
Comment (yahaa) [reply docs/storage.md]: Done.

**Comment by yeya24** [comment]
@yahaa Can you please fix the changelog?

**Comment by yahaa** [comment]
> @yahaa Can you please resolve the changelog conflicts?

Ok i have resolved the conflicts.😁

**Review by bwplotka — APPROVED** [review]
Amazing work! Clean, straightforward.

My instinct tells me we might have problem with `BOS` naming too - it might be bit cryptic, but since we allowed `COS` already, let's go for it 👍🏽 Thanks!

**Pull request merged**